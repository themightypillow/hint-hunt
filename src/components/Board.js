import React from "react";
import Letter from "./Letter";

class Board extends React.Component {
  state = {
    first: "",
    chain: [],
    classes: []
  };
  removeChild = (childSelector) => {
    let child = document.querySelector(childSelector);
    child.parentNode.removeChild(child);
  };
  removeChainClasses = () => {
    if(this.state.first !== "") {
      this.removeChild(`[data-coords="${this.state.chain[0]}"] div.${this.state.classes[0]}`);
      let length = this.state.chain.length;
      for(let i = 1; i < length; i++) {
        this.removeChild(`[data-coords="${this.state.chain[i]}"] div.${i === length - 1 ? this.state.classes[2] : this.state.classes[1]}`);
      }
    }
  };
  outlineWord = (first, firstClass, initValue, maxValue, middle, middleClass, last, lastClass) => {
    let chain = [first];
    this.props.addChild(`[data-coords="${first}"]`, firstClass);
    for(let i = initValue; i < maxValue; i++) {
      let evalMiddle = eval(middle);
      chain.push(evalMiddle);
      this.props.addChild(`[data-coords="${evalMiddle}"]`, middleClass);
    }
    chain.push(last);
    this.props.addChild(`[data-coords="${last}"]`, lastClass);
    this.setState(() => ({
      chain: chain,
      classes: [firstClass, middleClass, lastClass]
    }));
  };
  handleClick = (coords) => {
    this.props.addChild(`[data-coords="${coords}"]`, "Letter_circle");
    this.setState(() => ({ 
      first: coords,
      chain: [coords],
      classes: ["Letter_circle"]
    }));
  };
  handleMouseOver = (coords) => {
    this.removeChainClasses();
    if(this.state.first !== "") {
      if(this.state.first === coords) {
        this.props.addChild(`[data-coords="${this.state.first}"]`, "letter_circle");
        this.setState((prev) => ({ 
          chain: [prev.first],
          classes: ["letter_circle"]
        }));
      }
      else {
        let [row1, col1] = this.state.first.split("-");
        let [row2, col2] = coords.split("-");
        if(row1 === row2) {
          this.outlineWord(`${row1}-${Math.min(col1, col2)}`, 
                          "letter_left", 
                          Math.min(col1, col2) + 1, 
                          Math.max(col1, col2), 
                          "`${first.split('-')[0]}-${i}`", 
                          "letter_leftRight", 
                          `${row1}-${Math.max(col1, col2)}`, 
                          "letter_right"
          );
        }
        else if(col1 === col2) {
          this.outlineWord(`${Math.min(row1, row2)}-${col1}`, 
                          "letter_top", 
                          Math.min(row1, row2) + 1, 
                          Math.max(row1, row2), 
                          "`${i}-${first.split('-')[1]}`", 
                          "letter_topBottom", 
                          `${Math.max(row1, row2)}-${col1}`, 
                          "letter_bottom"
          );
        }
        else if((row2 - row1) === (col2 - col1)) {
          this.outlineWord(`${Math.min(row1, row2)}-${Math.min(col1, col2)}`, 
                          "letter_topLeft",
                          1,
                          Math.max(row1, row2) - Math.min(row1, row2),
                          "`${Number(first.split('-')[0]) + i}-${Number(first.split('-')[1]) + i}`",
                          "letter_backSlash",
                          `${Math.max(row1, row2)}-${Math.max(col1, col2)}`,
                          "letter_bottomRight"
          );
        }
        else if((row2 - row1) === -(col2 - col1)) {
          this.outlineWord(`${Math.min(row1, row2)}-${Math.max(col1, col2)}`,
                          "letter_topRight",
                          1,
                          Math.max(row1, row2) - Math.min(row1, row2),
                          "`${Number(first.split('-')[0]) + i}-${Number(first.split('-')[1]) - i}`",
                          "letter_forwardSlash",
                          `${Math.max(row1, row2)}-${Math.min(col1, col2)}`,
                          "letter_bottomLeft"
          );
        }
        else {
          this.props.addChild(`[data-coords="${this.state.first}"]`, "letter_circle");
          this.setState((prev) => ({ 
            chain: [prev.first],
            classes: ["letter_circle"]
          }));
        }
      }
    }
  };
  handleMouseUp = () => {
    let guess = this.state.chain.map((coords) => {
      let [row, column] = coords.split("-");
      let index = this.getIndex(Number(row), Number(column));
      return this.props.letters[index];
    }).join("");
    if(!this.props.handleGuess(guess)) {
      this.removeChainClasses();
    }
    this.setState((prev) => ({ 
      first: "",
      chain: [],
      classes: []
    }));
  };
  getCoords(index) {
    return [Math.floor(index / this.props.columns), index % this.props.columns];
  }
  getIndex(row, column) {
    return row * this.props.columns + column;
  }
  render() {
    return (
      <div className="container">
        <div className="board_grid">
          {
            this.props.letters.map((letter, index) => 
              <Letter 
                char={letter} 
                coords={`${this.getCoords(index)[0]}-${this.getCoords(index)[1]}`} 
                handleClick={this.handleClick}
                handleMouseOver={this.handleMouseOver}
                handleMouseUp={this.handleMouseUp}
                key={`${this.getCoords(index)[0]}-${this.getCoords(index)[1]}`}
              />
            )
          }
        </div>
      </div>
    );
  }
}

export default Board;