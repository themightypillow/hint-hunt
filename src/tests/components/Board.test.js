import React from "react";
import { shallow } from "enzyme";
import Board from "../../components/Board";

test("should render Board correctly", () => {
  const wrapper = shallow(<Board letters={["A"]} />);
  expect(wrapper).toMatchSnapshot();
});