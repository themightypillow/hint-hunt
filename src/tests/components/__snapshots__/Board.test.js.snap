// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Board correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Board
    letters={
      Array [
        "A",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="board_grid"
      >
        <Letter
          char="A"
          coords="NaN-NaN"
          handleClick={[Function]}
          handleMouseOver={[Function]}
          handleMouseUp={[Function]}
        />
      </div>,
      "className": "container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Letter
            char="A"
            coords="NaN-NaN"
            handleClick={[Function]}
            handleMouseOver={[Function]}
            handleMouseUp={[Function]}
          />,
        ],
        "className": "board_grid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "NaN-NaN",
          "nodeType": "function",
          "props": Object {
            "char": "A",
            "coords": "NaN-NaN",
            "handleClick": [Function],
            "handleMouseOver": [Function],
            "handleMouseUp": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="board_grid"
        >
          <Letter
            char="A"
            coords="NaN-NaN"
            handleClick={[Function]}
            handleMouseOver={[Function]}
            handleMouseUp={[Function]}
          />
        </div>,
        "className": "container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Letter
              char="A"
              coords="NaN-NaN"
              handleClick={[Function]}
              handleMouseOver={[Function]}
              handleMouseUp={[Function]}
            />,
          ],
          "className": "board_grid",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "NaN-NaN",
            "nodeType": "function",
            "props": Object {
              "char": "A",
              "coords": "NaN-NaN",
              "handleClick": [Function],
              "handleMouseOver": [Function],
              "handleMouseUp": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
